packages/security/src/types.ts

```diff
@@ -0,0 +1,183 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/security
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+/**
+ * A symbol for stringified id of security related objects
+ */
+export const securityId = Symbol('securityId');
+
+/**
+ * Represent a user, an application, or a device
+ */
+export interface Principal {
+  /**
+   * Name/id
+   */
+  [securityId]: string;
+
+  // tslint:disable-next-line:no-any
+  [attribute: string]: any;
+}
+
+export class TypedPrincipal implements Principal {
+  constructor(
+    public readonly principal: Principal,
+    public readonly type: string,
+  ) {}
+  get [securityId]() {
+    return `${this.type}:${this.principal[securityId]}`;
+  }
+}
+
+export interface UserProfile extends Principal {
+  // basic user information
+}
+
+export interface Organization extends Principal {}
+
+export interface Team extends Principal {}
+
+export interface ClientApplication extends Principal {}
+
+export interface Role extends Principal {}
+
+/**
+ * Security attributes used to authenticate the subject. Such credentials
+ * include passwords, Kerberos tickets, and public key certificates.
+ */
+export interface Credential {}
+
+/**
+ * `Permission` defines an action/access against a protected resource. It's
+ * the `what` for security.
+ *
+ * There are three levels of permissions
+ *
+ * - Resource level (Order, User)
+ * - Instance level (Order-0001, User-1001)
+ * - Property level (User-0001.email)
+ *
+ * @example
+ * - create a user (action: create, resource type: user)
+ * - read email of a user (action: read, resource property: user.email)
+ * - change email of a user (action: update, resource property: user.email)
+ * - cancel an order (action: delete, resource type: order)
+ */
+export class Permission {
+  /**
+   * Action or access of a protected resources, such as `read`, `create`,
+   * `update`, or `delete`
+   */
+  action: string;
+
+  /**
+   * Type of protected resource, such as `Order` or `Customer`
+   */
+  resourceType: string;
+  /**
+   * Property of a protected resource type/instance, such as `email`
+   */
+  resourceProperty?: string;
+  /**
+   * Identity of a protected resource instance, such as `order-0001` or
+   * `customer-101`
+   */
+  resourceId?: string;
+
+  get [securityId]() {
```

Comment:
Are we following any standard to generate the `[securityId]`?

My understanding is: we collect permission information(action, resourceType, etc...) from request, then generate `[securityId]` as a string, then sends it to the authorization server. And my question is: how do we know that the auth server understands the string?

Reply:
Please note that `securityId` is a `Symbol` - a unique key:

```ts
export const securityId = Symbol('securityId');
```

Reply:
@raymondfeng sorry I didn't put it clearly, I mean the value of key `[securityId]`, not the symbol itself. (And I like the idea of having the security id as a symbol)